src/fetch.ts

```diff
@@ -32,32 +43,151 @@ interface FetchConfig {
+  };
+}): JsonRpcMiddleware<JsonRpcParams, Json>;
+
+/**
+ * Creates middleware for sending a JSON-RPC request to the given RPC URL.
+ *
+ * @deprecated This overload is deprecated — please pass an `RpcService`
+ * instance from `@metamask/network-controller` instead.
+ * @param args - The arguments to this function.
+ * @param args.btoa - Generates a base64-encoded string from a binary string.
+ * @param args.fetch - The `fetch` function; expected to be equivalent to
+ * `window.fetch`.
+ * @param args.rpcUrl - The URL to send the request to.
+ * @param args.originHttpHeaderKey - If provided, the origin field for each
+ * JSON-RPC request will be attached to each outgoing fetch request under this
+ * header.
+ * @returns The fetch middleware.
+ */
+// eslint-disable-next-line @typescript-eslint/unified-signatures
+export function createFetchMiddleware(args: {
+  btoa: (stringToEncode: string) => string;
+  fetch: typeof fetch;
+  rpcUrl: string;
+  originHttpHeaderKey?: string;
+}): JsonRpcMiddleware<JsonRpcParams, Json>;
+
+export function createFetchMiddleware(
+  args:
+    | {
+        rpcService: AbstractRpcService;
+        options?: {
+          originHttpHeaderKey?: string;
+        };
+      }
+    | {
+        btoa: (stringToEncode: string) => string;
+        fetch: typeof fetch;
+        rpcUrl: string;
+        originHttpHeaderKey?: string;
+      },
+): JsonRpcMiddleware<JsonRpcParams, Json> {
+  if ('rpcService' in args) {
+    return createFetchMiddlewareWithRpcService(args);
+  }
+  return createFetchMiddlewareWithoutRpcService(args);
+}
+
+/**
+ * Creates middleware for sending a JSON-RPC request through the given RPC
+ * service.
+ *
+ * @param args - The arguments to this function.
+ * @param args.rpcService - The RPC service to use.
+ * @param args.options - Options.
+ * @param args.options.originHttpHeaderKey - If provided, the origin field for
+ * each JSON-RPC request will be attached to each outgoing fetch request under
+ * this header.
+ * @returns The fetch middleware.
+ */
+function createFetchMiddlewareWithRpcService({
+  rpcService,
+  options = {},
+}: {
+  rpcService: AbstractRpcService;
+  options?: {
+    originHttpHeaderKey?: string;
+  };
+}): JsonRpcMiddleware<JsonRpcParams, Json> {
+  return createAsyncMiddleware(
+    async (req: JsonRpcRequestWithOrigin<JsonRpcParams>, res) => {
+      const headers =
+        'originHttpHeaderKey' in options &&
+        options.originHttpHeaderKey !== undefined &&
+        req.origin !== undefined
+          ? { [options.originHttpHeaderKey]: req.origin }
+          : {};
+
+      const jsonRpcResponse = await rpcService.request(
+        {
+          id: req.id,
+          jsonrpc: req.jsonrpc,
+          method: req.method,
+          params: req.params,
+        },
+        {
+          headers,
+        },
+      );
+
+      // Discard the `id` and `jsonrpc` fields in the response body
+      // (the JSON-RPC engine will fill those in)
+
```

Comment:
nit: was this supposed to go a few lines down with this assignment?
```typescript
res.result = jsonRpcResponse.result;
```

Reply:
Yes, good catch. I have moved this closer in [12ec424](https://github.com/MetaMask/eth-json-rpc-middleware/pull/357/commits/12ec42419ce293c7370e140756f02d5678d57cab).